server/src/data_sources/builder.rs

```diff
@@ -1,79 +1,98 @@
-    types::{EdgeResult, EdgeSink, EdgeSource, EdgeToken},
+    types::{EdgeResult, EdgeSink, EdgeSource},
 };
 
 use super::{
     memory_provider::MemoryProvider, offline_provider::OfflineProvider,
-    redis_provider::RedisProvider,
+    redis_provider::RedisProvider, repository::DataSourceFacade,
 };
 
-pub type DataProviderPair = (Arc<RwLock<dyn EdgeSource>>, Arc<RwLock<dyn EdgeSink>>);
+pub type DataProviderPair = (Arc<dyn EdgeSource>, Arc<dyn EdgeSink>);
 
 pub struct RepositoryInfo {
-    pub source: Arc<RwLock<dyn EdgeSource>>,
+    pub source: Arc<dyn EdgeSource>,
     pub sink_info: Option<SinkInfo>,
 }
 
 pub struct SinkInfo {
-    pub sink: Arc<RwLock<dyn EdgeSink>>,
-    pub validated_send: mpsc::Sender<EdgeToken>,
-    pub validated_receive: mpsc::Receiver<EdgeToken>,
-    pub unvalidated_receive: mpsc::Receiver<EdgeToken>,
+    pub sink: Arc<dyn EdgeSink>,
     pub unleash_client: UnleashClient,
     pub token_validator: Arc<RwLock<TokenValidator>>,
     pub metrics_interval_seconds: u64,
 }
 
-fn build_offline(offline_args: OfflineArgs) -> EdgeResult<Arc<RwLock<dyn EdgeSource>>> {
+fn build_offline(offline_args: OfflineArgs) -> EdgeResult<Arc<dyn EdgeSource>> {
     let provider =
         OfflineProvider::instantiate_provider(offline_args.bootstrap_file, offline_args.tokens)?;
-    let provider = Arc::new(RwLock::new(provider));
-    Ok(provider)
+
+    let source: Arc<dyn EdgeSource> = Arc::new(provider);
+    Ok(source)
 }
 
-fn build_memory(features_refresh_interval_seconds: i64) -> EdgeResult<DataProviderPair> {
-    let data_source = Arc::new(RwLock::new(MemoryProvider::new(
-        features_refresh_interval_seconds,
-    )));
-    Ok((data_source.clone(), data_source))
+fn build_memory(features_refresh_interval_seconds: Duration) -> EdgeResult<DataProviderPair> {
+    let data_source = Arc::new(RwLock::new(MemoryProvider::new()));
+    let facade = Arc::new(DataSourceFacade {
+        features_refresh_interval: Some(features_refresh_interval_seconds),
+        token_source: data_source.clone(),
+        feature_source: data_source.clone(),
+        token_sink: data_source.clone(),
+        feature_sink: data_source,
+    });
+
+    let edge_source: Arc<dyn EdgeSource> = facade.clone();
+    let edge_sink: Arc<dyn EdgeSink> = facade;
+
+    Ok((edge_source, edge_sink))
 }
 
-fn build_redis(redis_url: String, _sender: Sender<EdgeToken>) -> EdgeResult<DataProviderPair> {
+fn build_redis(
+    redis_url: String,
+    features_refresh_interval_seconds: Duration,
+) -> EdgeResult<DataProviderPair> {
     let data_source = Arc::new(RwLock::new(RedisProvider::new(&redis_url)?));
-    Ok((data_source.clone(), data_source))
+    let facade = Arc::new(DataSourceFacade {
+        token_source: data_source.clone(),
+        feature_source: data_source.clone(),
+        token_sink: data_source.clone(),
+        feature_sink: data_source,
+        features_refresh_interval: Some(features_refresh_interval_seconds),
```

Comment:
This smells like a helper function, really only two arguments needed here, would be nice to not have to type it fully out each time. Not a blocker, just an improvement

Reply:
Yes this is a good suggestion. I'm not sure I'm content with this refactor yet, it's good enough for now but I want to do another pass to solve the threading issues. So this code might change during that restructure